internal/cache/rpc_cache.go

```diff
@@ -50,30 +62,45 @@ func (c *RPCCache) HandleRequest(chainName string, ttl time.Duration, reqBody js
 		result json.RawMessage
 	)
 
+	labels := prometheus.Labels{"request": reqBody.Method, "chain": chainName}
+
+	// Counts requests in flight. If it's spiking that means that a lot of requests are for the same key.
+	// If there's a too many requests in flight, and redis latency is high it means that the cache is down.
+	c.metricsContainer.CacheRequestsInFlight.With(labels).Inc()
+
+	start := time.Now()
+
 	// Even if the cache is down, redis-cache will route to the origin
 	// properly without returning an error.
 	// Do() is executed on cache misses or if the cache is down.
 	// Item.Value is still set even on cache miss and the request to origin succeeds.
 	// Once() uses request coalescing (single in-flight request to origin even if there are
 	// multiple identical incoming requests), which means returned errors
 	// could be coming from other goroutines.
-	err := c.Once(&cache.Item{
+	err := c.cache.Once(&cache.Item{
 		Key:   c.CreateRequestKey(chainName, reqBody),
 		Value: &result,
 		TTL:   ttl,
 		Do: func(*cache.Item) (interface{}, error) {
 			cached = false
+
+			// Capturing the duration from when the request to redis was initiated to when we detect a cache miss.
+			c.metricsContainer.CacheQueryCacheMissDuration.With(labels).Observe(time.Since(start).Seconds())
 			respBody, err := originFunc()
 			if err != nil {
 				return nil, err
 			}
+			c.metricsContainer.CacheMiss.With(labels).Inc()
 			return &respBody.Result, nil
 		},
 	})
 
 	if err != nil {
 		return nil, cached, err
 	}
+	if !cached {
+		c.metricsContainer.CacheWriteDuration.With(labels).Observe(float64(time.Since(start)))
```

Comment:
```suggestion
		writeDuration := time.Since(start) - cacheLookupDuration - originDuration
		c.metricsContainer.CacheWriteDuration.With(labels).Observe(writeDuration.Seconds())
```